Setup script and file structure with placeholders (#7)
badgerloop-software/BTF: README.md

# Badgerloop Testing Framework

## Setup
This repository is designed for and validated on a Raspberry Pi 4 running Ubuntu 22.04 Jammy Jellyfish. The Pi is expected to have the following packages:
- python3 (Should be included by default on Ubuntu 22.04)
- python3-venv (sudo apt update && sudo apt install python3-venv)
- pigpiod (install using instructions [here](https://abyz.me.uk/rpi/pigpio/download.html))

> Additionally, you will need a testConfig file to indicate what mocks are needed. 

Reviewer: ```suggestion
```
I think these lines can be removed since it's specified in the "Running Tests" section

Reviewer: Nvm. This is fine

## Running Tests

> 1. Run the setup script to create/update the python environment and fetch all required packages (./setup.sh </folder/to/create/in>)

Reviewer: ```suggestion
1. Run the setup script to create/update the python environment and fetch all required packages (`./setup.sh </folder/to/create/in>`)
```

> 2. Run the test script. You will need a [configuration file](#writing-the-configuration-file) to tell the runner what devices to mock and how. (`python3 testRunner.py </path/to/testConfig.extension>`)

Reviewer: ```suggestion
2. Run the test script. You will need two [configuration files](#writing-the-configuration-files) to tell the runner what devices to mock and how. (`python3 testRunner.py </path/to/testConfig.yml> </path/to/setupConfig.yml>`)
```

> ## Writing the Configuration File

Reviewer: ```suggestion
## Writing the Configuration Files
```

> TBD

Reviewer: Structures are defined in `tests.yml` and `setup.yml` in `embedded-mbed`, so this can probably just point to those files

Reviewer: Pointed to the embedded-mbed README, as the configuration file structures will be added there

# Badgerloop Testing Framework

## Setup
This repository is designed for and validated on a Raspberry Pi 4 running Ubuntu 22.04 Jammy Jellyfish. The Pi is expected to have the following packages:
- python3 (Should be included by default on Ubuntu 22.04)
- python3-venv (sudo apt update && sudo apt install python3-venv)

> - pigpiod (sudo apt update && sudo apt install pigpiod)

Reviewer: There were 2 packages that came up in a search for packages matching pigpiod: `libpigpiod-if-dev` and `libpigpiod-if2-1`. It seems like there are other packages built from the same source: https://packages.ubuntu.com/source/jammy/pigpio. We should probably install `pigpio-tools` as well. It looks like `python3-pigpio` is just the Python module specified in requirements.txt and can be ignored.

Author: Looks like they actually recommend installing directly from the repo https://abyz.me.uk/rpi/pigpio/download.html. Will update pending tests with the Pi tomorrow (9/5/2023)

Reviewer: Sounds good. Yeah, feel free to update it to install from the source

Additionally, you will need a testConfig file to indicate what mocks are needed. 

## Running Tests
1. Run the setup script to create the python environment and fetch all required packages (./setup.sh </folder/to/create/in>)

> 2. Activate the python environment (source </folder/you/specified/above>)

Reviewer: Can we change `</folder/to/create/in>` to `</path/to/venv/parent>` or something along those lines (could use `/path/to/new/virtual/environment` to match the Python docs)? Also, step 2 should be taken care of when running `./setup.sh`, so that can be merged into step 1. Maybe something like:
```suggestion
1. Run the setup script to create and activate the python environment and fetch all required packages (`./setup.sh </path/to/venv/parent>`)
```

Author: Ok with the argument change. Currently the script is deactivating the venv at the end. I also think step 2 should be separated out since for manual use, this gives context on how to run tests later on without rerunning the entire setup script.

Reviewer: What if we don't deactivate the venv at the end of the script then? It seems kind of redundant to deactivate the environment and then reactivate it. Also, by activating the environment using `setup.sh` each time, we're ensuring that the environment being used is up-to-date. Even if we wanted to avoid running `setup.sh` every time, we could still use the activated venv after setting it up for the first time. We could add a printout at the end of `setup.sh` to say something like:
```
Currently running in the virtual environment at ${venv_path}. This environment is set up for BTF usage.

If any Python packages need to be installed or updated, exit the virtual environment, update
requirements.txt accordingly, and rerun setup.sh.

You can exit this virtual environment by running `deactivate`
```

> 3. Run the test script. You will need a [configuration file](#writing-the-configuration-file) to tell the runner what devices to mock and how. (python3 testRunner.py </path/to/testConfig/file>)

Reviewer: ```suggestion
3. Run the test script. You will need a [configuration file](#writing-the-configuration-file) to tell the runner what devices to mock and how. (`python3 testRunner.py </path/to/testConfig>`)
```
nit. Just adding code styling and stopping at the file name (if you want to). It would be nice to add a file extension to `/path/to/testConfig` to be explicit about the expected file type, but I think we should save that until we can add instructions for "Writing the Configuration File." We could keep the `/file` if you want and just replace it with a file extension once we know for sure what we're using.

Author: Styling is a good idea, will try to finalize the extension as part of work session tomorrow (9/5/2023)

> ## Writing the Configuration File

Reviewer: ```suggestion

## Writing the Configuration File
```
nit
